File: packages/template-go/functions/string.go
package functions

import (
	"strings"

	"github.com/polyspec/template/value"
)

func asciiUpper(text string) string {
	b := []byte(text)
	for i, c := range b {
		if c >= 'a' && c <= 'z' {
			b[i] = c - 32
		}
	}
	return string(b)
}

func asciiLower(text string) string {
	b := []byte(text)
	for i, c := range b {
		if c >= 'A' && c <= 'Z' {
			b[i] = c + 32
		}
	}
	return string(b)
}

var stringFunctions = map[string]BuiltIn{
	"upper": {1, 1, func(args []value.Value, _ Context) (value.Value, error) {
		text, err := argString(args[0], "upper")
		if err != nil {
			return nil, err
		}
		return asciiUpper(text), nil
	}},
	"lower": {1, 1, func(args []value.Value, _ Context) (value.Value, error) {
		text, err := argString(args[0], "lower")
		if err != nil {
			return nil, err
		}
		return asciiLower(text), nil
	}},
	"trim": {1, 2, func(args []value.Value, _ Context) (value.Value, error) {
		text, err := argString(args[0], "trim")
		if err != nil {
			return nil, err
		}
		chars, err := optString(args, 1, " \t\r\n", "trim")
		if err != nil {
			return nil, err
		}
		return strings.Trim(text, chars), nil
	}},
	"replace": {3, 3, func(args []value.Value, _ Context) (value.Value, error) {
		text, err := argString(args[0], "replace")
		if err != nil {
			return nil, err
		}
		from, err := argString(args[1], "replace")
		if err != nil {
			return nil, err
		}
		to, err := argString(args[2], "replace")
		if err != nil {
			return nil, err
		}
		if from == "" {
			return text, nil
		}
		return strings.ReplaceAll(text, from, to), nil
	}},
	"split": {2, 2, func(args []value.Value, _ Context) (value.Value, error) {
		text, err := argString(args[0], "split")
		if err != nil {
			return nil, err
		}
		sep, err := argString(args[1], "split")
		if err != nil {
			return nil, err
		}
		if sep == "" {
			return nil, typeError("split requires a non-empty separator")
		}
		parts := strings.Split(text, sep)
		list := make(value.List, len(parts))
		for i, part := range parts {
			list[i] = part
		}
		return list, nil
	}},
	"truncate": {2, 3, func(args []value.Value, _ Context) (value.Value, error) {
		text, err := argString(args[0], "truncate")
		if err != nil {
			return nil, err
		}
		limit, err := argInteger(args[1])
		if err != nil {
			return nil, err
		}
		suffix, err := optString(args, 2, "...", "truncate")
		if err != nil {
			return nil, err
		}
		runes := []rune(text)
		if len(runes) > limit {
			if limit < 0 {
				limit = 0
			}
			return string(runes[:limit]) + suffix, nil
		}
		return text, nil
	}},
	"contains": {2, 2, func(args []value.Value, _ Context) (value.Value, error) {
		if haystack, ok := value.TextOf(args[0]); ok {
			needle, ok := value.TextOf(args[1])
			if !ok {
				return nil, typeError("contains requires a string needle for a string haystack")
			}
			return strings.Contains(haystack, needle), nil
		}
		if list, ok := args[0].(value.List); ok {
			for _, item := range list {
				if value.LooseEquals(item, args[1]) {
					return true, nil
				}
			}
			return false, nil
		}
		return nil, typeError("contains requires a string or a list")
	}},
	"starts_with": {2, 2, func(args []value.Value, _ Context) (value.Value, error) {
		s, err := argString(args[0], "starts_with")
		if err != nil {
			return nil, err
		}
		p, err := argString(args[1], "starts_with")
		if err != nil {
			return nil, err
		}
		return strings.HasPrefix(s, p), nil
	}},
	"ends_with": {2, 2, func(args []value.Value, _ Context) (value.Value, error) {
		s, err := argString(args[0], "ends_with")
		if err != nil {
			return nil, err
		}
		p, err := argString(args[1], "ends_with")
		if err != nil {
			return nil, err
		}
		return strings.HasSuffix(s, p), nil
	}},
	"length": {1, 1, func(args []value.Value, _ Context) (value.Value, error) {
		switch x := args[0].(type) {
		case nil:
			return float64(0), nil
		case value.List:
			return float64(len(x)), nil
		case *value.OrderedMap:
			return float64(x.Len()), nil
		}
		if text, ok := value.TextOf(args[0]); ok {
			return float64(value.CodePointLength(text)), nil
		}
		return nil, typeError("length requires a string, list, map or null")
	}},
}

File: packages/template-ts/src/functions/string.ts
// upper, lower, trim, replace, split, truncate, contains, starts_with, ends_with (FUN-12, FUN-13).
import { codePointLength, codePoints, looseEquals, isString, textOf, type Value } from '../value/value.js';
import { argInteger, argString, typeError, type BuiltIn } from './helpers.js';

function asciiUpper(text: string): string {
  return text.replace(/[a-z]/g, char => String.fromCharCode(char.charCodeAt(0) - 32));
}

function asciiLower(text: string): string {
  return text.replace(/[A-Z]/g, char => String.fromCharCode(char.charCodeAt(0) + 32));
}

function trimChars(text: string, chars: string): string {
  const set = new Set(codePoints(chars));
  const points = codePoints(text);
  let start = 0;
  let end = points.length;
  while (start < end && set.has(points[start] as string)) start++;
  while (end > start && set.has(points[end - 1] as string)) end--;
  return points.slice(start, end).join('');
}

export const stringFunctions: Record<string, BuiltIn> = {
  upper: { min: 1, max: 1, call: ([s]) => asciiUpper(argString(s as Value, 'upper')) },
  lower: { min: 1, max: 1, call: ([s]) => asciiLower(argString(s as Value, 'lower')) },
  trim: {
    min: 1,
    max: 2,
    call: ([s, chars]) => trimChars(argString(s as Value, 'trim'), chars === undefined ? ' \t\r\n' : argString(chars, 'trim')),
  },
  replace: {
    min: 3,
    max: 3,
    call: ([s, from, to]) => {
      const text = argString(s as Value, 'replace');
      const search = argString(from as Value, 'replace');
      const replacement = argString(to as Value, 'replace');
      return search === '' ? text : text.split(search).join(replacement);
    },
  },
  split: {
    min: 2,
    max: 2,
    call: ([s, sep]) => {
      const text = argString(s as Value, 'split');
      const separator = argString(sep as Value, 'split');
      if (separator === '') throw typeError('split requires a non-empty separator');
      return text.split(separator);
    },
  },
  truncate: {
    min: 2,
    max: 3,
    call: ([s, n, suffix]) => {
      const text = argString(s as Value, 'truncate');
      const limit = argInteger(n as Value, 'truncate');
      const tail = suffix === undefined ? '...' : argString(suffix, 'truncate');
      const points = codePoints(text);
      return points.length > limit ? points.slice(0, Math.max(0, limit)).join('') + tail : text;
    },
  },
  contains: {
    min: 2,
    max: 2,
    call: ([h, n]) => {
      const haystack = h as Value;
      const needle = n as Value;
      if (isString(haystack)) {
        if (!isString(needle)) throw typeError('contains requires a string needle for a string haystack');
        return textOf(haystack).includes(textOf(needle));
      }
      if (Array.isArray(haystack)) return haystack.some(item => looseEquals(item, needle));
      throw typeError('contains requires a string or a list');
    },
  },
  starts_with: { min: 2, max: 2, call: ([s, p]) => argString(s as Value, 'starts_with').startsWith(argString(p as Value, 'starts_with')) },
  ends_with: { min: 2, max: 2, call: ([s, p]) => argString(s as Value, 'ends_with').endsWith(argString(p as Value, 'ends_with')) },
  length: {
    min: 1,
    max: 1,
    call: ([v]) => {
      const value = v as Value;
      if (value === null) return 0;
      if (isString(value)) return codePointLength(textOf(value));
      if (Array.isArray(value)) return value.length;
      if (value instanceof Map) return value.size;
      throw typeError('length requires a string, list, map or null');
    },
  },
};
